Add a generic script for deploying WMAgent inside a virtual environment.

todor-ivanov opened the pull request:
Fixes #11617
#### Status
ready

#### Description
With the current script we provide a method for implementing the same set of steps as done during the deployment process inside the WMAgent docker image, but inside a well contained virtual environment. The end result should be a runtime environment equal to the one inside the Docker image. 
**NOTE:** All Os dependencies need to be resolved in advance  
  
**NOTE:** Here follow some instructions on how to deploy and run the agent:
```
./wmagent-venv-deploy.sh  -d /data/WMAgent.venv3 -t 2.2.1 -y
source /data/WMAgent.venv3/bin/activate
source $WMA_ENV_FILE
/data/WMAgent.venv3/run.sh
manage status
```


#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
https://github.com/dmwm/CMSKubernetes/pull/1364 

#### External dependencies / deployment changes
https://github.com/dmwm/CMSKubernetes/pull/1364

Once the above PR in CMSkubernetes is merged we should proceed and fix the following lines in this script before merging it:
 **TODO:** 
To change 
```
deployRepo=https://github.com/todor-ivanov/CMSKubernetes.git
deployBranch=Fix_WMAgentPyPi_DcokerImage 
```
to 
```
deployRepo=https://github.com/dmwm/CMSKubernetes.git 
deployBranch=master
```

todor-ivanov committed:
Add a generic script for deploying wmagent inside a virtual environment.

Change mode to 755 from wmagent-venv-deploy.sh

Fix python executable link

Remove setupDependencies for run from source mode && Fix outdated Rucio links && Fix help bool flags

Substitute export command with _addWMCoreVenvVar function && Add X509 variables to the environment.

Removing serviceList && Removing setupRucio && Cloning CMSKubernetes from dmwm

Catchup with WMAgent init changes

Rename wmagent-venv-deploy to deploy-wmagent-venv.sh

ReviewComments && Fix missing tag param && Remove _realPath && Remove sudo commands

amaltaro commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+    venvPath=$(_realPath $venvPath)
+    # TODO: To be changed to point to master at https://github.com/dmwm/CMSKubernetes.git
+    #       once we have it all merged at CMSKubernetes
+    deployRepo=https://github.com/todor-ivanov/CMSKubernetes.git
```
Todor, for the future, please make a clear note (TODO) in the PR description that such lines need to be corrected before merging.

todor-ivanov replied:
The PR description has been updated.

amaltaro replied:
Should this be updated now?

todor-ivanov replied:
done

amaltaro commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+    rm -rf $venvPath/CMSKubernetes
+
+    # Install the requested WMA_TAG.
+    ${WMA_ROOT_DIR}/install.sh -v ${WMA_TAG}
```
When we install WMAgent through this `install.sh` script, which under the hood will use `pip install`, it is supposed to bring in all of the python dependencies with the package.

This makes me wonder if there is any actual need for the function `setupDependencies`? I guess the answer is "it depends". If we want to have all of the WMCore dependencies, then the best is to install all those packages. But given that this script is being provided to the T0 to install WMAgent from PyPi, I would say that we should not install all of the WMCore dependencies, but only those that are actually needed by T0 agent.

todor-ivanov replied:
You are correct. 
It would be needed only for `run from source` mode. So if  in this mode we also deploy the service package, indeed it won't be needed, but if we don't, there is the need for it.

todor-ivanov replied:
fixed  now

amaltaro commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+    # chown -R ${WMA_USER}:${WMA_GID} ${WMA_ROOT_DIR}
+
+    # Add all environment variables to the virtEnv hooks
+    # _addWMCoreVenvVar X509_USER_CERT ${wmAuthPath}/dmwm-service-cert.pem
```
These grid certificate names are valid for central services, but I think in the agent we have different certificate names.

todor-ivanov replied:
this is not needed here at all. All those for the agent come from the $WMA_ENV_FILE. I'll remove that line.

todor-ivanov replied:
I actually simply added them to the environment - it makes things better.

amaltaro reviewed (changes requested):
@todor-ivanov as we discussed this week, I'd rather have someone from T0 giving this script a try and providing us with feedback. 
Nonetheless, I left a few relevant comments along the code that you might want to address before collecting feedback.

todor-ivanov commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+                                  [-g <wmcore_config_repository>] [-d wmagent_path] [-l <service_list>] [-i <pypi_index>]
+                                  [-h <help>]
+
+      -s  <run_from_source>          Bool flag to setup run from source [Default: false]
```
Agreed. Similarly I removed those abbreviations from the rest of the boolean flags as well.

todor-ivanov commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+    exit 1
+}
+
+_realPath(){
```
In the past I was simply using `realpath` unix command, but Valentin mentioned this was missing for MAC machines, so I developed this auxiliary function and extensively tested it. It has proved itself handy to prevent many mistakes.

todor-ivanov replied:
done

todor-ivanov commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+      -y  <assume yes>               Bool flag to assume 'Yes' to all deployment questions.
+      -r  <wmcore_source_repository> WMCore source repository [Default: git://github.com/dmwm/wmcore.git"]
+      -b  <wmcore_source_branch>     WMCore source branch [Default: master]
+      -t  <wmcore_tag>               WMCore tag to be used for this deployment [Default: None]
```
Yes, it was intended for that exact purpose, but until we use the script for deploying just the `wmagent` package I can remove the option.

amaltaro replied:
I think this script should be made specific to WMAgent, thus removing the `[-l <service_list>]` option.

todor-ivanov replied:
done

todor-ivanov commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,760 @@
+[common]
+[client]
+rucio_host = http://cmsrucio-int.cern.ch
+auth_host = https://cmsrucio-auth-int.cern.ch
```
fixed

todor-ivanov commented:
Hi @amaltaro I thinks I have addressed all your comments. Couls youplease take another loook
As of the T0 side, @germanfgv is already testing the PR. He was reporting some OS dependencies like: 
```
ERROR: No matching distribution found for Jinja2~=3.1.2
```
I will give him the list of all the OS packages that  may be needed for this setup.

amaltaro commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,672 @@
+    _pipUpgradeVenv
+}
+
+setupRucio(){
```
Don't we get the `rucio-clients` package from the requirements of WMAgent?
On what concerns the rucio.cfg, isn't it shipped with the pypi wmagent package as well?

todor-ivanov replied:
It was actually even not called from the main function at all. Removed now.

amaltaro commented on deploy/wmagent-venv-deploy.sh:
```diff
@@ -0,0 +1,672 @@
+}
+
+setupIpython(){
+    # Helper function to install Ipython during manual installation, it is skipped by default.
```
It is actually installed by default, not skipped.

todor-ivanov replied:
comment fixed

amaltaro reviewed (changes requested):
@todor-ivanov please see a few extra comments inline. Once the relevant changes are made, we could then ask German to have another look and try this out.

todor-ivanov commented:
hi @amaltaro please I have addressed all your comments and tested the script. Please  take another look
@germanfgv  please update your clone before your next test.

amaltaro commented on deploy/deploy-wmagent-venv.sh:
```diff
@@ -0,0 +1,628 @@
+help(){
+    echo -e $*
+    cat <<EOF
+    Usage: wmagent-venv-deploy.sh [-s] [-n] [-v] [-y]
```
these examples/documentation does not match the script name

todor-ivanov replied:
done

amaltaro reviewed (changes requested):
Todor, I left a comment in the code, but I didn't review most of the code.
It's still not clear to me whether we will be using this script for the agent, but if you are using it or prefer to get it merged, let's get it done.

todor-ivanov commented:
hi @LinaresToine  there was a dedicated effort from both of us lately in the form of hands on meetings and redeployment &&  readjustments in replay agents etc. There are still things to be done of course, like the oracle functionalities, which are about to come later during the completion of the meta issue about the full containerization of the WMAgents: https://github.com/dmwm/WMCore/issues/11314 && https://github.com/dmwm/WMCore/issues/11720 , but they won't require any further deployments  or tests related to the functionalities provided with this script here.  

I'd like to  close the issue https://github.com/dmwm/WMCore/issues/11617, which is basically for confirming that the so provided script here make things to happen and we are capable of having a generic WMAgent deployment with it. Which is something we did prove with you during our meetings. If you agree with me, we should ask from @amaltaro  for final approval and we should merge the script and close the issue.

todor-ivanov commented:
@amaltaro  I did address all your comments and squashed all the commits, If you find it ok now, lets merge this.

amaltaro reviewed (approved):
Thanks Todor.
